### ISSUE TITLE
Networking port dns_name reset to "" when running terraform apply on existing cluster with v0.17.0

### ISSUE THREAD

**Issue by tobgu** [terraform-providers/terraform-provider-openstack#742]
Hi there,

Since upgrading to 0.16.0 -> 0.17.0 terraform wants to reset the `dns_name` of ports to the empty string.

I guess this happens because of https://github.com/terraform-providers/terraform-provider-openstack/commit/80c17a90108ad39b9f9329cb06004b0d65a27620 but wouldn't it make sense to be backwards compatible and use `name` as `dns_name` if no `dns_name` was provided in the configuration and the current `dns_name` is the same as `name`? Or something similar...

### Terraform Version
Terraform v0.11.13
+ provider.null v2.1.1
+ provider.openstack v1.17.0
+ provider.template v2.1.1
+ provider.tls v1.2.0

### Affected Resource(s)
Please list the resources as a list, for example:
- openstack_networking_port_v2

### Expected Behavior
No updates required when running `terraform plan` after initial `terraform apply`.

### Actual Behavior
Terraform/openstack provider now wants to clear the DNS name of all ports

### Steps to Reproduce
1. Create port without `dns_name` set but with `name` set then `terraform apply`
2.  `openstack port show <port id>`, dns_name will be set to the value of `name`.
3. `terraform plan`, terraform now wants to clear `dns_name` of all ports as per printout below.
```
Terraform will perform the following actions:

  ~ openstack_networking_port_v2.host[0]
      dns_name: "host1" => ""

  ~ openstack_networking_port_v2.host[1]
      dns_name: "host2" => ""

  ~ openstack_networking_port_v2.host[2]
      dns_name: "host3" => ""
```

**Comment by kayrus** [comment]
Sorry for this. Indeed, we don't have tests for this use case. I assume that a simple patch will fix the issue:

```diff
			"dns_name": {
￼ 				Type:     schema.TypeString,
￼ 				Optional: true,
+￼ 				Computed: true,
￼ 			},
```

Let's wait for maintainers' opinion on this.

**Comment by jtopjian** [comment]
> I assume that a simple patch will fix the issue

I think so, yes.

> Probably it makes sense to add Computed: true for new dns field in other resources.

Yes, I think so, too.

In environments that don't have the DNS extension enabled, this will cause the user to see computed fields in their plan, but they _should_ end up being set to an empty string. I think that's OK.

> no, they're not the same. May I ask you how exactly did you create these ports? Who did assign the "dns_name" on these ports?

They're not the same, but the Neutron DNS extension will set `dns_name` automatically in certain situations. See https://docs.openstack.org/neutron/queens/admin/config-dns-int.html at the bottom ("The following is an example of an instance creation, showing how its hostname populates the dns_name attribute of the allocated port")

**Comment by kayrus** [comment]
> They're not the same, but the Neutron DNS extension will set dns_name automatically in certain situations. See https://docs.openstack.org/neutron/queens/admin/config-dns-int.html at the bottom ("The following is an example of an instance creation, showing how its hostname populates the dns_name attribute of the allocated port")

In that situation port is created automatically by the Openstack backend and it is not managed by Terraform. The issue is still unclear, I'd wait for the info on how the ports were created and managed.

**Comment by jtopjian** [comment]
There are two options:

1. We make the field computed which would end up clobbering the automatically-set `dns_name`. This would cater to users who want to set `dns_name`.

2. Make the field computed-only. This would cater to users who prefer obtaining the automatically-set `dns_name`.

Option 3 is a combination of the two, but I can't think of a clean way for how it would work (yet).

I agree - getting some more use-cases is definitely the best for a long-term solution. But depending on how many others are affected by this, we should consider some kind of intermediary fix if possible.

**Comment by jtopjian** [comment]
Thinking about this more, I think Option 1 is the way to go. If a user sets `dns_name`, then they are explicitly opting into overriding any `dns_name` that was set by the compute service.

**Comment by tobgu** [comment]
Option 1 seems reasonable to me. What additional information is it that you would like to have?

While it wouldn't be a great deal to explicitly set the `dns_name` to the same value as the name for us I figured that the backwards incompatibility was unnecessary.

**Comment by kayrus** [comment]
@tobgu I still don't think it is a backward compatibility issue. I assume three scenarios:
1) you created VMs, then imported ports, which were automatically created, into terraform, adjusted terraform scripts and forgot about this. In this case, we should not apply the `compute:true` patch, but probably create a doc note, or custom state migration logic.
2) you created these ports using terraform, then somehow assigned them to compute instances and nova/neutron modified `dns_name`. My tests didn't show such a behavior. Port's `dns_name` is modified only when port is created automatically by the backend for the compute instance.
3) your openstack backend runs custom code, which automatically modifies `dns_name` according to the port `name`.

Two last cases could be a vanilla code behavior, but I'm not aware about it. In this case we have to find this code, confirm its behavior and apply the `compute:true` patch.

Please don't get me wrong. It is easy to add a `compute:true` field, however I'd prefer to identify the root cause, because I have some doubts about the openstack code. If @jtopjian approves once again, that `compute: true` is the way to resolve an issue, I'll do it.

So far, as a workaround I suggest you to add corresponding `dns_name` values to your port definitions, and terraform won't try to modify them.

P.S. Please try to recreate the env from scratch and let me know whether the issue persists.

@jtopjian , adding the `compute: true` can lead to a behavior, when user omits the `dns_name` field and assumes that it will be removed, but terraform will keep it.

**Comment by jtopjian** [comment]
> I still don't think it is a backward compatibility issue

I agree with this. There is no issue with backwards compatibility here. This is an issue with a new feature that _might_ be clobbering existing behavior. It's a bug.

> you created these ports using terraform, then somehow assigned them to compute instances and nova/neutron modified dns_name. My tests didn't show such a behavior. 

I am able to replicate this behavior in a standard OpenStack environment:

Variation 1:

In this variation, I set `dns_name` to `jtopjian` and then create an instance using that port. The `dns_name` is then overwritten to `jttest`.

```
openstack port create --network <uuid> --dns-name jtopjian jttest
openstack server create ... --nic port-id=<port-uuid> jttest
openstack port show <port-uuid>
```

Variation 2:

I create a port with no `dns_name` and then create a server using that port. `dns_name` is then set.

```
openstack port create --network <uuid> jttest
openstack server create ... --nic port-id=<port-uuid> jttest
openstack port show <port-uuid>
```

> So far, as a workaround I suggest you to add corresponding dns_name values to your port definitions, and terraform won't try to modify them.

This is a temporary workaround, sure, but it is definitely not the behavior we want.

>  adding the compute: true can lead to a behavior, when user omits the dns_name field and assumes that it will be removed, but terraform will keep it.

This is because `dns_name` is truly a computed field. 

Clearing the name might require some research or an additional argument of `clear_dns_name` (probably not the best option), but it is not a reason to prevent `dns_name` from being computed.

**Comment by jtopjian** [comment]
@kayrus At the moment, the only solution I can think of is to set this field to be computed.

This might end up being a case where the user must use their discretion on how to use the field: by default, an OpenStack environment will set `dns_name` automatically. If the user wants to override that behavior, that's something they will need to manage themselves. If there's a way to clear `dns_name`, we can amend that in the future.

However, I appreciate the push-back you've given. If you feel strongly that there's an alternative fix, I'm definitely open to waiting until you can look into it. But we should get some kind of bug fix merged sooner than later.

**Comment by jtopjian** [comment]
@tobgu This has been fixed in #748. We'll get a new release out early next week.

Thank you very much for reporting this and for your patience.

**Comment by tobgu** [comment]
Thanks for fixing it, much appreciated!
